subcommunicator abort

yfguo opened the pull request:
All abort on a communicator that is not MPI_COMM_WORLD and has at least two processes would internally the newly added `MPIDIG_comm_abort` function. The function will send abort message to all the processes in the communicator over active message. A new message type `MPIDI_CH4U_COMM_ABORT` is added for this purpose. If a send to a particular rank returns error, the function will ignore it and keep going. After completing the send of all abort message, the caller exits. All the other processes will exit upon receiving the abort message.

wesbland commented on test/mpi/ft/testlist:
```diff
@@ -20,3 +20,4 @@ agree_shrink 4 env=MPIR_CVAR_ENABLE_FT=1 mpiexecarg=-disable-auto-cleanup result
 revoke_shrink 4 env=MPIR_CVAR_ENABLE_FT=1 mpiexecarg=-disable-auto-cleanup resultTest=TestStatusNoErrors strict=false timeLimit=10 xfail=ticket2198
 nbccoll 4 env=MPIR_CVAR_ENABLE_FT=1 mpiexecarg=-disable-auto-cleanup resultTest=TestStatusNoErrors strict=false timeLimit=10 xfail=ticket2237
 failure_ack 3 env=MPIR_CVAR_ENABLE_FT=1 mpiexecarg=-disable-auto-cleanup resultTest=TestStatusNoErrors strict=false timeLimit=10 xfail=ticket2237
+subcomm 4 mpiexecarg=-disable-auto-cleanup timeLimit=10 strict=false resultTest=TestStatus
```
Should this test go in the `errors` directory instead of the `ft` directory? This isn't exactly FT and doesn't require the FT runtime flags to be set I think.

yfguo replied:
Good point. Let me change it.

yfguo commented:
@wesbland I think two processes calling abort on the same subcomm is fine. This is no need of matching a recv for the abort message. All processes should receive it as an unexpected message, and do `MPL_exit` in the send_target_cb.

wesbland commented:
Ok, it might be worth adding that test case just to be sure.

yfguo commented:
Added.

yfguo commented:
test:jenkin/ch4/most

yfguo commented:
test:jenkins/ch4/most

yfguo commented:
I have fixed the testing. I have also added the initial support for abort on intercomm (abort remote group).

yfguo commented:
test:jenkins/ch4/all

yfguo commented:
test:jenkins/ch4/all

yfguo commented:
test:jenkins/ch4/ucx
test:jenkins/ch4/ofi

yfguo commented:
test:jenkins/ch4/ucx

yfguo commented:
test:jenkins/ch4/all

yfguo commented:
test:jenkins/ch4/all

yfguo commented:
test:jenkins/ch4/all

raffenet commented on src/mpid/ch4/src/ch4r_callbacks.h:
```diff
@@ -634,4 +634,43 @@ static inline int MPIDI_send_long_ack_target_msg_cb(int handler_id, void *am_hdr
+#define FUNCNAME MPIDI_comm_abort_target_msg_cb
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+static inline int MPIDI_comm_abort_target_msg_cb(int handler_id, void *am_hdr,
```
We should make this a `void` function. It will never return.

yfguo replied:
That might be a problem because of the registration function in `generic/mpidig_init.c` takes `MPIDIG_am_target_msg_cb` type parameter which explicitly says the return value be `int`.

yfguo replied:
On a second thought, let me try it first. Since return type is not part of the function signature, it might work.

raffenet replied:
Ah, ok. Well let's see how it goes. We should do whatever makes the compiler happy in that case.

yfguo replied:
The code is running fine. But compiler gives a warning of incompatible pointer type for that. I would leave it as `int` just for the cleanness.

raffenet commented on src/mpid/ch4/generic/mpidig_globals.c:
```diff
@@ -11,6 +11,60 @@
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDIG_COMM_ABORT);
+
+  fn_exit:
+    /* MPL_exit(exit_code); */
```
Can we just call this instead of sending an abort message to self? That way we won't automatically leak the send request on `MPL_exit`.

raffenet replied:
And in that case, this function should also be `void`.

raffenet replied:
@yfguo to be clear, I still prefer making this function `void` and removing the self send at the end in favor of calling `MPL_exit`.

yfguo replied:
I think it is OK to call `MPL_exit` here instead of doing a self send for intracomm. The problem was with intercomm abort which should abort the remote group and the local rank who calls the abort should return success. But I just realized what I did is wrong in having the `MPIDIG_comm_abort` always do a self send.

raffenet replied:
OK, can you fix it up and send a new version? I'm going to hold off on my proposal for now and work on other PRs.

yfguo replied:
Sure. Will ping you when it is updated.

yfguo commented:
test:jenkins/ch4/ucx
test:jenkins/ch4/ofi

yfguo commented:
@raffenet I tested it on Jenkins last night and the results come back as clean (but I squashed the commit and the results are gone). I have changed the function signature to `void` and fix the exit handling.

raffenet commented:
OK, I will take a look.

raffenet commented on src/mpid/ch4/generic/mpidig_globals.c:
```diff
@@ -11,6 +11,49 @@
+    am_hdr.tag = exit_code;
+    am_hdr.context_id = comm->context_id + MPIR_CONTEXT_INTRA_PT2PT;
+
+    for (dest = 0; dest < comm->size; dest++) {
```
This won't compile, though it gets changed later so it's not noticeable. It should be either `comm->local_size` or `MPIR_Comm_size(comm)`.

raffenet replied:
Maybe we should just squash this with the next commit that gets rid of this line. I don't know if there's benefit to having separate commits for the intra/intercomm versions. And just to reiterate, this function probably needs to go back to returning an `int` for the intercomm version.

raffenet commented on src/mpid/ch4/src/ch4r_callbacks.h:
```diff
@@ -634,4 +634,43 @@ static inline int MPIDI_send_long_ack_target_msg_cb(int handler_id, void *am_hdr
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_COMM_ABORT_TARGET_MSG_CB);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_COMM_ABORT_TARGET_MSG_CB);
+
+  fn_exit:
```
No need for `fn_exit, fn_fail` labels here. The compiler will just complain.

yfguo replied:
Wouldn't the `MPIR_ERR_CHKANDSTMT` above need these labels? Or should we just remove all return code checkings from this function?

raffenet replied:
I'm not seeing `MPIR_ERR_CHKANDSTMT` in this function.

yfguo replied:
Oh, I was looking at the wrong function. Fixing it right away.

yfguo replied:
Done. Also got rid of the unused `mpi_errno`.

raffenet commented on src/mpid/ch4/generic/mpidig_globals.c:
```diff
@@ -53,7 +60,10 @@ void MPIDIG_comm_abort(MPIR_Comm * comm, int exit_code)
+    if (comm->comm_kind == MPIR_COMM_KIND__INTRACOMM)
+        MPL_exit(exit_code);
+    else
+        return MPI_SUCCESS;
```
I mentioned this yesterday, but now I get why you made this function return `int`. We probably should change it back from `void` if there is the option of not exiting the calling process. Sorry for the churn.

raffenet commented on src/mpid/ch4/src/ch4_globals.c:
```diff
@@ -128,7 +128,7 @@ int MPID_Abort(MPIR_Comm * comm, int mpi_errno, int exit_code, const char *error
 #endif
     }
-    return 0;
+    return MPI_SUCCESS;
```
Unrelated change.

raffenet commented:
Looks good. I made a small tweak when signing off. Going to test one more time. The only other thought I had is are these tests in the correct directory? I dunno, but we can figure that out later. Thanks for fixing all the comments.
test:jenkins/ch4/most

yfguo commented:
test:jenkins/ch4/most
